src/Compilers/CSharp/Portable/CodeGen/Optimizer.cs

```diff
@@ -2071,8 +2082,12 @@ public override BoundNode VisitAssignmentOperator(BoundAssignmentOperator node)
             // fake visiting of left
             _nodeCounter += 1;
 
+            Debug.Assert(!_isStackScheduledRefAssignment);
+            _isStackScheduledRefAssignment = node.IsRef;
+
             // visit right
             var right = (BoundExpression)Visit(node.Right);
+            _isStackScheduledRefAssignment = false;
```

Comment:
>false [](http://example.com/codeflow?start=45&length=5)

Should this reset to the previous value rather than `false`? (Is there a test we can write where that makes a difference?)

Reply:
No, I asserted that it was false above as well. `Visit` should be saving and resetting the value unless the thing being visited is an indirection or conditional operator, so we should never be observing a non-false value here.

Reply:
Ok, thanks.

Could we save and restore the value anyway, for clarity, and to avoid a bug later if there are additional cases where the save and restore in `Visit` is not sufficient?

Reply:
I suppose I could, but I'll keep the assertion anyway. If we come across a case like that we definitely need to understand the semantics more completely.